File: src/repositories/vendor-price-rate.repository.ts
import {Collection, ObjectId} from "mongodb";
import type {VendorPriceRate} from "@/types/vendor/vendor_price_rate";
import {logger} from "@/utils/logger";
import {getVendorPriceRateCollection} from "@/repositories/collections/core.collections";

export class VendorPriceRateRepository {
  private getCollection(companyId: string): Collection<VendorPriceRate> {
    return getVendorPriceRateCollection(companyId);
  }

  /**
   * Find price rates for multiple vendors
   */
  async findByVendorIds(companyId: string, vendorIds: ObjectId[]): Promise<VendorPriceRate[]> {
    try {
      const collection = this.getCollection(companyId);
      return await collection
        .find({vendorId: {$in: vendorIds}})
        .toArray();
    } catch (error) {
      logger.error("Failed to fetch vendor price rates", error);
      throw error;
    }
  }

  /**
   * Bulk upsert price rates — insert or update by vendorId
   */
  async bulkUpsert(companyId: string, rates: VendorPriceRate[]): Promise<void> {
    try {
      if (rates.length === 0) {
        return;
      }

      const collection = this.getCollection(companyId);
      const operations = rates.map((rate) => ({
        updateOne: {
          filter: {vendorId: new ObjectId(rate.vendorId)},
          update: {$set: {rate: rate.rate}},
          upsert: true,
        },
      }));

      await collection.bulkWrite(operations);
      logger.info("Vendor price rates bulk upserted", {count: rates.length});
    } catch (error) {
      logger.error("Failed to bulk upsert vendor price rates", error);
      throw error;
    }
  }
}

File: fisoft/src/main/kotlin/repository/VendorPriceRateRepository.kt
package com.ilyasdemirkiran.repository

import com.ilyasdemirkiran.config.Database
import com.ilyasdemirkiran.types.vendors.VendorPriceRate
import com.mongodb.client.model.Filters
import com.mongodb.client.model.UpdateOptions
import com.mongodb.client.model.Updates
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.flow.toList
import org.bson.types.ObjectId

class VendorPriceRateRepository {
    private fun getCollection(companyId: String): MongoCollection<VendorPriceRate> {
        return Database.getDatabaseForCompany(companyId).getCollection<VendorPriceRate>("vendor_price_rates")
    }

    suspend fun findByVendorIds(companyId: String, vendorIds: List<String>): List<VendorPriceRate> {
        val objIds = vendorIds.map { try { ObjectId(it) } catch (e: Exception) { it } }
        return getCollection(companyId).find(Filters.`in`("vendorId", objIds)).toList()
    }

    suspend fun upsertRate(companyId: String, vendorId: String, rate: Double): VendorPriceRate {
        val vObjId = try { ObjectId(vendorId) } catch (e: Exception) { vendorId }
        val filter = Filters.eq("vendorId", vObjId)
        val update = Updates.combine(
            Updates.set("rate", rate),
            Updates.set("updatedAt", java.time.Instant.now().toString())
        )
        val options = UpdateOptions().upsert(true)
        getCollection(companyId).updateOne(filter, update, options)

        return VendorPriceRate(
            companyId = companyId,
            vendorId = vendorId,
            rate = rate,
            updatedAt = java.time.Instant.now().toString()
        )
    }
}
